core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -182,6 +190,10 @@ private[spark] class ExecutorAllocationManager(
   // ResourceProfile id to Host to possible task running on it, used for executor placement.
   private var rpIdToHostToLocalTaskCount: Map[Int, Map[String, Int]] = Map.empty
 
+  private var adjustMaxNeed = minNumExecutors;
+
+  private val excludeSpeculativeNodes = new mutable.HashSet[String]()
```

Comment:
At any given point of time, there can be multiple stages concurrently running - and for each stage, different sets of executor(s)/hosts might require to be speculated : some might be slow for memory reasons, while others might be slow for IO reasons,
Given this, we cannot assume a given executor host is bad for all stages.

Instead, track this per stage attempt.
(This applies to other fields in this class as well - the change in design should fix them all).